src/dir/glob.cr

```diff
@@ -1,266 +1,369 @@
         end
       end
-      nested_path = nested_path[0...lastidx]
+    end
 
-      recursion_depth -= depth if recursion_depth != Int32::MAX
-      dir = File.join(dir, nested_path)
-      if !nested_path.empty? && nested_path[0] == File::SEPARATOR
-        nested_path = nested_path[1..-1]
-      end
+    private def compile(pattern, patterns = [] of Array(PatternType))
+      reader = Char::Reader.new(pattern)
 
-      regex = glob2regex(ptrn)
+      lbrace = nil
+      rbrace = nil
+      alt_start = nil
 
-      scandir(dir, nested_path, regex, 0, recursion_depth, wants_dir) do |path|
-        if starts_with_path_separator
-          yield File::SEPARATOR + path
-        else
-          yield path
+      alternatives = [] of String
+
+      nest = 0
+      while char = reader.current_char
+        case char
+        when Char::ZERO
+          break
+        when '{'
+          lbrace = reader.pos if nest == 0
+          nest += 1
+        when '}'
+          nest -= 1
+
+          if nest == 0
+            rbrace = reader.pos
+            start = (alt_start || lbrace).not_nil! + 1
+            alternatives << pattern.byte_slice(start, reader.pos - start)
+            break
+          end
+        when ','
+          if nest == 1
+            start = (alt_start || lbrace).not_nil! + 1
+            alternatives << pattern.byte_slice(start, reader.pos - start)
+            alt_start = reader.pos
+          end
         end
+
+        reader.next_char
       end
-    end
-  end
 
-  private def self.glob2regex(pattern)
-    if pattern.size == 0 || pattern == File::SEPARATOR
-      raise ArgumentError.new "Empty glob pattern"
-    end
+      if lbrace && rbrace
+        front = pattern.byte_slice(0, lbrace)
+        back = pattern.byte_slice(rbrace + 1)
 
-    # characters which are escapable by a backslash in a glob pattern;
-    # Windows paths must have double backslashes:
-    escapable = {'?', '{', '}', '*', ',', '\\'}
-    # characters which must be escaped in a PCRE regex:
-    escaped = {'$', '(', ')', '+', '.', '[', '^', '|', '/'}
+        alternatives.each do |alt|
+          brace_pattern = [front, alt, back].join
 
-    last_is_file_separator = false
+          compile brace_pattern, patterns
+        end
+      else
+        patterns << single_compile pattern
+      end
 
-    regex_pattern = String.build do |str|
-      str << "\\A"
-      idx = 0
-      nest = 0
+      patterns
+    end
 
-      idx = 1 if pattern[0] == File::SEPARATOR
-      size = pattern.size
+    private def single_compile(glob)
+      list = [] of PatternType
+      return list if glob.empty?
 
-      while idx < size
-        char = pattern[idx]
+      parts = glob.split('/', remove_empty: true)
 
-        if last_is_file_separator && char == File::SEPARATOR
-          idx += 1
-          next
+      if glob.ends_with?('/')
+        list << DirectoriesOnly.new
+      else
+        file = parts.pop
+        if /^[a-zA-Z0-9._]+$/.match(file)
```

Comment:
I thought regex wasn't used 😄

Reply:
Not for the globbing itself ^^

Reply:
I think not using a regex is not that hard in this case. We can use this method, I think:

```crystal
private def constant_entry?(file)
  file.each_char do |char|
    return false if char == '*' || char == '?'
  end

  true
end
```

I _think_ we can use that here and a bit below (in the other regex). I'm not sure why the regexes differ. But if you need a difference you can always write two methods with similar but different checks. In my benchmarks this improves things a little. The other bottleneck is `File.join`, I'll try to optimize that for common cases (here a common case is joining 2 pieces).

Reply:
Yes, that seems to be fine. But we'll need to update the special characters when others get added.

I'd also like this:
```crystal
SPECIAL = {'?', '*'}
return false if SPECIAL.includes?(c)
```
But it is a bit slower than the individual comparisons...